src/interactive.jl

```diff
@@ -0,0 +1,179 @@
+    convert_input(::Type{P}, ::Type{T}, s::AbstractString) -> T
+
+Convert the user input `s` into an instance of `T`.
+A default implementation of `T(s)` exists.
+"""
+convert_input(::Type, T::Type{<:Real}, s::AbstractString) = parse(T, s)
+convert_input(::Type, T::Type, s::AbstractString) = T(s)
+
+function convert_input(P::Type, ::Type{Union{T, Nothing}}, s::AbstractString) where T
+    # This is kind of sketchy because technically, there might be some other input
+    # whose value we want to instantiate with the string "nothing",
+    # but I think that would be a pretty rare occurrence.
+    # If that really happens, they can just override this method.
+    return s == "nothing" ? nothing : convert_input(P, T, s)
+end
+
+function convert_input(::Type, ::Type{Bool}, s::AbstractString)
+    s = lowercase(s)
+    return if startswith(s, "t") || startswith(s, "y")
+        true
+    elseif startswith(s, "f") || startswith(s, "n")
+        false
+    else
+        throw(ArgumentError("Unrecognized boolean response"))
+    end
+end
+
+function convert_input(P::Type, T::Type{<:Vector}, s::AbstractString)
+    startswith(s, "[") && endswith(s, "]") && (s = s[2:end-1])
+    xs = map(strip, split(s, ","))
+    return map(x -> convert_input(P, eltype(T), x), xs)
+end
+
+"""
+    prompt(::Type{P}, ::Type{T}, ::Val{name::Symbol}) -> Any
+
+Prompts for an input of type `T` for field `name` of plugin type `P`.
+Implement this method to customize particular fields of particular types.
+"""
+prompt(P::Type, T::Type, name::Symbol) = prompt(P, T, Val(name))
+
+# The trailing `nothing` is a hack for `fallback_prompt` to use, ignore it.
+function prompt(P::Type, ::Type{T}, ::Val{name}, ::Nothing=nothing) where {T, name}
+    tips = join([T; input_tips(T); "default=$(repr(defaultkw(P, name)))"], ", ")
+    default = defaultkw(P, name)
+    print(pretty_message("Enter value for '$name' ($tips): "))
+    input = strip(readline())
+    return if isempty(input)
+        default
+    else
+        try
+            # Working around what appears to be a bug in Julia 1.0:
+            # #145#issuecomment-623049535
+            if VERSION < v"1.1" && T isa Union && Nothing <: T
+                if input == "nothing"
+                    nothing
+                else
+                    convert_input(P, T.a === Nothing ? T.b : T.a, input)
+                end
+            else
+                convert_input(P, T, input)
+            end
+        catch ex
+            ex isa InterruptException && rethrow()
+            @warn "Invalid input" ex
+            prompt(P, T, name)
+        end
+    end
+end
+
+# Compute all the concrete subtypes of T.
+concretes_rec(T::Type) = isabstracttype(T) ? vcat(map(concretes_rec, subtypes(T))...) : Any[T]
+concretes(T::Type) = sort!(concretes_rec(T); by=nameof)
+
+# Compute name => type pairs for T's interactive options.
+function interactive_pairs(T::Type)
+    pairs = collect(map(name -> name => fieldtype(T, name), fieldnames(T)))
```

Comment:
I think this is clearer:
```suggestion
    pairs = collect(Pair.(fieldnames(T), fieldtypes(T)))
```
Also turns out to be the fastest:
```julia
julia> T = Rational
Rational

julia> @btime collect(map(name -> name => fieldtype($T, name), fieldnames($T)))
  15.374 μs (8 allocations: 368 bytes)
2-element Array{Pair{Symbol,DataType},1}:
 :num => Integer
 :den => Integer

julia> @btime collect(Pair.(fieldnames($T), fieldtypes($T)))
  3.055 μs (11 allocations: 464 bytes)
2-element Array{Pair{Symbol,DataType},1}:
 :num => Integer
 :den => Integer

julia> @btime [fieldname($T, i) => fieldtype($T, i) for i in 1:fieldcount($T)]
  16.317 μs (9 allocations: 416 bytes)
2-element Array{Pair{Symbol,DataType},1}:
 :num => Integer
 :den => Integer
```